### ISSUE TITLE
Unify span context when components are nested

### ISSUE THREAD

**Issue by binarylogic** [timberio/vector#4454]
According to https://github.com/timberio/vector/pull/4181#issuecomment-700391110, the metrics labels from nested components will not reflect the parent component.

For example, the `humio_logs` sink wraps the `splunk_hec` sink. The `splunk_hec` sink [defines a `SplunkEventSent`](https://github.com/timberio/vector/blob/3507ed2d00d95011e0f9a19b24da02a9014caa54/src/internal_events/splunk_hec.rs#L14) which emits the `events_processed` metric. This metric will include labels that identify the `splunk_hec` sink instead of the `humio_logs` sink. To a user this is confusing, since they defined a `humio_logs` sink, not a `splunk_hec` sink.

**Comment by MOZGIII** [comment]
I think I have an idea for the solution! When passing the execution to the wrapped components, we can just add another `span!` around the execution block.

This will allow us to either pass custom `component_type`/`component_name`, or pass some additional fields - for instance, `wrapped = true` to indicate that the executing transform is wrapped with some component and that it's `processed_events` sould be excluded when evaluating the overall sum of the processed events.

**Comment by leebenson** [comment]
@MOZGIII, to confirm -- does you see this issue (or #4181) being all that is required to help unblock the format of `vector top` proposed by #3211:

```
$ vector top
ID              KIND       TYPE         THRPT   I/O      LATENCY    ERRORS
my-file-source  source     file         5.2k    10.6MiB  10.2ns     251
my-json-parser  transform  json_parser  5.0k    -        1.2ms      523
my-s3-sink      sink       s3           4.5k    5.2MiB   10.2s      12
```

Or do you think we'll need a follow-up issue to address separating metrics by `component_name`?

**Comment by MOZGIII** [comment]
In my view, this is not enough to make the `top` command work (i.e. show reliable data). At the same time, it's not a blocker.

To elaborate - this issue focuses on a particular known problem of metrics, however it does not guarantee that, once resolved, the data provided by our internal metrics is sane/accurate for the needs of the top command. I feel we need a proper in-depth audit of metrics to find and resolve other discrepancies and edge cases.

At the same time, the implementation of the top command should not be stalled by this, on the other hand, the efforts to make top work and show accurate data may be the primary driver of fixes to the metrics subsystem. I believe, when you start with integrating with the internal metrics system (aggregating metrics for display at top), some more subtle problems will become more obvious. Or maybe there aren't any, I can't really tell without digging deeper.

**Comment by leebenson** [comment]
Gotcha, thanks @MOZGIII.

So the issue with `vector top` currently is that there isn't _any_ way to associate a metric with an individual source, per the layout @binarylogic proposed in #3230.

We can get top-level stats - like total events processed, total errors, total bytes processed, etc - but AFAICT, we can't identify which sources generated them. This prevents being able to show things like:

```
$ vector top
ID              KIND       TYPE         THRPT   I/O      LATENCY    ERRORS
my-file-source  source     file         5.2k    10.6MiB  10.2ns     251
my-json-parser  transform  json_parser  5.0k    -        1.2ms      523
my-s3-sink      sink       s3           4.5k    5.2MiB   10.2s      12
```

Instead, we'd be able to show something akin to:

```
$ vector top
METRIC              DATA
uptime              23.5 seconds
events processed    12,927
bytes processed     2,948,294
```

... which can still update in real-time, but perhaps isn't very useful or what we had in mind for `vector top`.

What I'd like to be able to do is pull metrics out based on an individual resource (source/sink), so instead of 'events processed' being a global stat, it relates to a given piece of topology. It seemed like this / #4181 might be work toward that goal, but I may have been misunderstood the scope.

**Comment by MOZGIII** [comment]
Ah, https://github.com/timberio/vector/pull/3888 has that covered already, and it's been merged for a while.

**Comment by leebenson** [comment]
Ah nice, thanks, I totally missed that. Let me play with it a bit locally and see if it solves what I'm trying to do. I'll revert back if I run into any questions - thanks in advance!

**Comment by MOZGIII** [comment]
@leebenson I've just merged #4181, you might want to pull in the changes.

**Comment by ktff** [comment]
From I gathered, it seams the issue is of one event incrementing a metric multiple times because the component is comprised of multiple smaller components where each of them emit the the same metric. While the issue mentioned in the opening comment has been resolved with #4181. @MOZGIII is this correct?

### The short
Basically this comment https://github.com/timberio/vector/pull/4181#issuecomment-707392168

The fix:
> Move `event_processed` internal events to topology wrappers instead of
    components themselves to avoid double counting or incorrect tagging (likely
    within `impl Transform for TransformFn` for now)

### Reasoning

The metrics can be separated into two categories depending on how they answer the question `If we have a component that reuses two other components, and we rewrite it so that it doesn't use them, would we still emit the metric the same number of times`:
* `Yes` - these metrics can't be double counted.
* `No` - these metrics can be double counted.

 From a rough scan of current metrics, only two metrics have `No` as an answer:
* `processed_bytes_total` 
* `events_processed_total`

Where `processed_bytes_total` isn't an issue so long as sources aren't combined with sinks. With that,  only `events_processed_total` needs to be considered.

**Comment by MOZGIII** [comment]
> While the issue mentioned in the opening comment has been resolved with #4181. @MOZGIII is this correct?
@ktff this is correct, the example in this issue is indeed solved by the https://github.com/timberio/vector/pull/4181.

The problem is that https://github.com/timberio/vector/pull/4181 causes us to merge the double-counted metrics together, making them indistinguishable. Before metrics would have different `component_type` label values (as they were specified literally), but now, if the source, for instance, reuses a transform internally, it will bump the `processed_bytes_total` counter twice *with exactly the same label values*.

Regarding the proposed topology wrappers fix - that would sort of make the problem go away effectively. Do we really only have just those two events that suffer from this? If so - this would be a great fix, but I think there are also things like `processing_errors_total` and etc, which are much more meaningful if they preserve the information about which one of the subcomponents was the origin of the metric. I.e. for a `json_parser` nested within `kubernetes_logs` source having labels `processing_errors_total{origin => json_parser, component_type => kubernetes_logs}` is more meaningful than just `processing_errors_total{component_type => kubernetes_logs}` (`component_type` being added from a `span`). I think there are more of such metrics that should be generically aggregatable, yet will benefit if we add more specific info for the cases that need it.

**Comment by ktff** [comment]
> processing_errors_total and etc, which are much more meaningful if they preserve the information about which one of the subcomponents was the origin of the metric.

@MOZGIII that's true, it would be meaningful, but they aren't broken as are the double counted one so I think we should address them separately. For kinds like `processing_errors_total` it makes sense to enrich their `span!` with additional labels, like you mentioned in the above comment, like `sub_component_type` although I'm not sure how much of an performance impact would that have, to have to enter and exit additional span for every event.

**Comment by MOZGIII** [comment]
> enter and exit additional span for every event

Spans are quite expensive, we might want to bench this before going forward.

**Comment by ktff** [comment]
So this turned out more nuanced than I anticipated. While we could raise emitting `events_processed_total` to topology level for sources and transforms, it would count different things for sinks. Where it now counts sent events, it would count events entering sinks. This mostly differs for metric sinks, but would also differ for those sinks with batching, or with some kind of filtering.

Also `processed_bytes_total` also turned out to be double counted when source and sink reuse other source and sink respectively. 

While we could raise `processed_bytes_total` for sources and transforms, it would introduce some complexity in that that sinks need to emit this while sources and transforms don't.

Instead, we can add label `origin`, that @MOZGIII mentioned [here](https://github.com/timberio/vector/issues/4454#issuecomment-716249927), with component type which would be defined at metric emit site. This would solve the issue, give us more meaning and would solve the loss of meaning for events like `processing_errors_total`. With this the real `processed_bytes_total` and `events_processed_total` counters for a component would have `origin` and `component_type` be equal. 

While `origin` is more verbose, it would be more difficult to contribute/maintain components with the original solution.

**Comment by ktff** [comment]
The above comment is quite mouthful so here is a conciser proposition that would replace [the older one](https://github.com/timberio/vector/issues/4454#issuecomment-716235511). 

### Proposal
Add label `origin` to internal metrics which would correspond to `type` of the actual component/code emitting it. This label would be added on the emit site. With that we can:
- Avoid double counting of metrics. (Implementation would under unique `name` pick one `origin` whose counter would be used for the whole component)
- Retain fine grained information about metrics, while being able to get aggregates.
- Shared code that isn't a component can have `origin` of it's own. 
- More easily add new metrics to reused components and shared code. They can just be added wherever we want without regards to what else is emitting them.

cc. @MOZGIII @leebenson

**Comment by MOZGIII** [comment]
Yep, sounds good! We might want to rethink the names (`component_name` / `origin / etc), but the core approach of adding this data-point solves the issue.
